// @angular/catalyst/dist/test_component.js
import { Component, Input, NgModule } from '@angular/core';
import { MatIconModule } from '@angular/material/icon';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/icon";
export class AotTestComponent {
    constructor() {
        this.name = 'World';
    }
    click() { }
}
AotTestComponent.ɵfac = function AotTestComponent_Factory(t) { return new (t || AotTestComponent)(); };
AotTestComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AotTestComponent, selectors: [["aot-test-component"]], inputs: { name: "name" }, decls: 5, vars: 1, consts: [[3, "click"]], template: function AotTestComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "p", 0);
        i0.ɵɵlistener("click", function AotTestComponent_Template_p_click_0_listener() { return ctx.click(); });
        i0.ɵɵtext(1);
        i0.ɵɵelementStart(2, "mat-icon");
        i0.ɵɵtext(3, "face");
        i0.ɵɵelementEnd()();
        i0.ɵɵtext(4, "\n");
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate1("Hello ", ctx.name, " ");
    } }, directives: [i1.MatIcon], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AotTestComponent, [{
        type: Component,
        args: [{ preserveWhitespaces: true, selector: 'aot-test-component', template: "<p (click)=\"click()\">Hello {{name}} <mat-icon>face</mat-icon></p>\n" }]
    }], null, { name: [{
            type: Input
        }] }); })();
export class AotParentComponent {
}
AotParentComponent.ɵfac = function AotParentComponent_Factory(t) { return new (t || AotParentComponent)(); };
AotParentComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AotParentComponent, selectors: [["aot-parent-test-component"]], decls: 1, vars: 0, consts: [["name", "World"]], template: function AotParentComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelement(0, "aot-test-component", 0);
    } }, directives: [AotTestComponent], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AotParentComponent, [{
        type: Component,
        args: [{
                preserveWhitespaces: true,
                selector: 'aot-parent-test-component',
                template: `<aot-test-component name="World"></aot-test-component>`,
            }]
    }], null, null); })();
export class AotTestModule {
}
AotTestModule.ɵfac = function AotTestModule_Factory(t) { return new (t || AotTestModule)(); };
AotTestModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AotTestModule });
AotTestModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[MatIconModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AotTestModule, [{
        type: NgModule,
        args: [{
                imports: [MatIconModule],
                declarations: [AotParentComponent, AotTestComponent],
                exports: [AotTestComponent],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AotTestModule, { declarations: [AotParentComponent, AotTestComponent], imports: [MatIconModule], exports: [AotTestComponent] }); })();
