// @angular/catalyst/dist/examples/example_module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ExampleComponent } from './example_component';
import * as i0 from "@angular/core";
export class ExampleModule {
}
ExampleModule.ɵfac = function ExampleModule_Factory(t) { return new (t || ExampleModule)(); };
ExampleModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: ExampleModule });
ExampleModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(ExampleModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ExampleComponent,
                ],
                imports: [
                    CommonModule,
                ],
                exports: [
                    ExampleComponent,
                ],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(ExampleModule, { declarations: [ExampleComponent], imports: [CommonModule], exports: [ExampleComponent] }); })();
