// @angular/catalyst/dist/examples/directive/example_directive.js
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Replaces span textContent with text provided in content input.
 *
 * Also contains property with value of textContent capitalized.
 */
export class ExampleDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.content = '';
        this.contentCapitalized = '';
    }
    ngOnChanges() {
        this.elementRef.nativeElement.textContent = this.content;
        this.contentCapitalized = this.content.toUpperCase();
    }
}
ExampleDirective.ɵfac = function ExampleDirective_Factory(t) { return new (t || ExampleDirective)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
ExampleDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: ExampleDirective, selectors: [["span", "replaceContent", ""]], inputs: { content: "content" }, features: [i0.ɵɵNgOnChangesFeature] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(ExampleDirective, [{
        type: Directive,
        args: [{
                selector: 'span[replaceContent]',
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { content: [{
            type: Input
        }] }); })();
