// @angular/catalyst/dist/dist/shared-angular-catalyst/integration_test.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import '@angular/compiler';
import { ChangeDetectionStrategy, Component, CUSTOM_ELEMENTS_SCHEMA, EventEmitter, Inject, Injectable, InjectionToken, Input, NgModule, NgZone, Output, RendererFactory2, ViewChild } from '@angular/core';
import { MatIconModule } from '@angular/material/icon';
import { afterEach, afterEachLegacy, beforeEach, beforeEachLegacy, bootstrap, bootstrapTemplate, describe, describeWithDate, destroyTestComponent, expect, flush, get, getDebugEl, getDebugEls, getEl, getEls, hasDebugEl, hasEl, it, itLegacy, markForCheckAndFlush, now, setupModule, tick, trigger } from './index.js';
import { from, Observable } from 'rxjs';
import { debounceTime, delay } from 'rxjs/operators';
import { AotParentComponent, AotTestComponent, AotTestModule } from './test_component.js';
const _global = (typeof window === 'undefined' ? global : window);
describe('catalyst', () => {
    let state = 0;
    beforeAll(() => {
        // beforeAll should be executed without any errors.
        expect(true).toBe(true);
    });
    afterAll(() => {
        // afterAll should be executed without any errors.
        expect(true).toBe(true);
    });
    // Use the unpatched afterEach to make sure the patched one executes.
    // TODO(b/186802882): Remove any after Jasmine upgrade is in. Element
    // implicitly has an 'any' type because expression of type '"afterEach"' can't
    // be used to index type 'Env'.
    jasmine.getEnv()['afterEach'](() => {
        expect(state).toBe(3);
    });
    beforeEach(() => {
        state = 1;
    });
    it('executes patched test methods', () => {
        expect(state).toBe(1);
        state = 2;
    });
    afterEach(() => {
        expect(state).toBe(2);
        state = 3;
    });
});
describe('catalyst', () => {
    it('expects nothing', () => {
        expect().nothing();
    });
});
describe('catalyst', () => {
    let comp;
    let root;
    beforeEach(() => {
        setupModule({ imports: [TestModule] });
        comp = bootstrap(TestComponent);
        // getEl() is the root app element.
        root = getEl();
    });
    _global.it('APIs throw exceptions outside the catalyst \'it\'', () => {
        expect(() => bootstrap(TestComponent))
            .toThrowError(`Put bootstrap() inside beforeEach, beforeAll or it. Remember to ` +
            `import 'it', 'beforeEach', 'afterEach' etc. from the catalyst` +
            ` module.`);
        expect(() => get(NgZone))
            .toThrowError(`Put get() inside beforeEach, beforeAll or it. Remember to ` +
            `import 'it', 'beforeEach', 'afterEach' etc. from the catalyst` +
            ` module.`);
    });
    it('runs in Catalyst fakeAsync Zone', () => {
        expect(Zone.current.get('Catalyst')).toBe(true);
    });
    it('bootstraps', () => {
        expect(root.textContent).toEqual('Hello World face');
    });
    it('get() retrieves injected object', () => {
        expect(get(NgZone)).not.toBe(null);
    });
    it('runs change detection automatically on event', () => {
        // Click the text.
        getEl('p').click();
        expect(root.textContent).toEqual('Hello Clicked face');
    });
    it('runs change detection after a flush', () => {
        // flush() reflects changes made directly in component instance to DOM.
        comp.name = 'Earth';
        flush();
        expect(root.textContent).toEqual('Hello Earth face');
    });
    it('runs change detection after an async now', () => {
        // now will wait for the async method to finish before running the
        // change detection that will reflect the state to the DOM.
        now(() => __awaiter(void 0, void 0, void 0, function* () {
            yield null;
            comp.name = 'Mars';
        }));
        expect(root.textContent).toEqual('Hello Mars face');
    });
    it('runs change detection after a synchronous now', () => {
        now(() => {
            comp.name = 'Mars';
        });
        expect(root.textContent).toEqual('Hello Mars face');
    });
    it('runs test in fakeAsync zone by default', () => {
        const asyncMethod = () => __awaiter(void 0, void 0, void 0, function* () {
            yield null;
            comp.name = 'Mars';
        });
        asyncMethod(); // This queues the async task.
        expect(root.textContent).toEqual('Hello World face');
        flush(); // This processes any queued async tasks.
        expect(root.textContent).toEqual('Hello Mars face');
    });
    it('works with async/await test method', () => __awaiter(void 0, void 0, void 0, function* () {
        yield null;
        comp.name = 'Mars';
        flush();
        expect(root.textContent).toEqual('Hello Mars face');
        yield null;
        comp.name = 'Venus';
        flush();
        expect(root.textContent).toEqual('Hello Venus face');
    }));
    itLegacy('supports legacy tests by not flushing after `it()`', () => __awaiter(void 0, void 0, void 0, function* () {
        expect().nothing(); // Suppress warning of 0 assertions in this test.
        yield new Promise((resolve) => {
            setTimeout(() => {
                resolve();
            });
        });
        // Not called because the above Promise is not implicitly flushed.
        fail();
    }));
    // This complicated setup is needed to check if `it.broken` causes a warning
    // to be printed at the end. The actual expectation is in afterEach.
    describe('with console.warn mocked', () => {
        let spy;
        beforeEach(() => {
            spy = spyOn(console, 'warn');
        });
        // Deprecated version intentionally used here.
        // tslint:disable-next-line:deprecation
        it.broken('prints a warning when returned promise not resolved', () => {
            return new Promise(() => { });
        });
        afterEach(() => {
            expect(spy.calls.count()).toEqual(1);
            expect(spy.calls.mostRecent().args[0])
                .toMatch(/Test returned a promise that didn't resolve within /);
        });
    });
    it('strict works with async/await', () => __awaiter(void 0, void 0, void 0, function* () {
        yield new Promise((resolve) => {
            setTimeout(() => {
                resolve();
            });
        });
        flush();
        expect(true).toBe(true);
    }));
    describe('supports legacy tests by not flushing after `beforeEach()`', () => {
        beforeEachLegacy(() => __awaiter(void 0, void 0, void 0, function* () {
            yield new Promise((resolve) => {
                setTimeout(() => {
                    resolve();
                });
            });
            // Not called because the above Promise is not implicitly flushed.
            fail();
        }));
        itLegacy('empty test so `beforeEach()` will run', () => {
            // Suppress warning of 0 assertions in this test.
            expect(true).toBe(true);
        });
    });
    describe('supports legacy tests by not flushing after `afterEach()`', () => {
        afterEachLegacy(() => __awaiter(void 0, void 0, void 0, function* () {
            yield new Promise((resolve) => {
                setTimeout(() => {
                    resolve();
                });
            });
            // Not called because the above Promise is not implicitly flushed.
            fail();
        }));
        itLegacy('empty test so `afterEach()` will run', () => {
            // Suppress warning of 0 assertions in this test.
            expect(true).toBe(true);
        });
    });
    it('flushes after `it()`', () => __awaiter(void 0, void 0, void 0, function* () {
        yield new Promise((resolve) => {
            setTimeout(() => {
                resolve();
            });
        });
        // Called because the above Promise is implicitly flushed.
        expect(true).toBe(true);
    }));
    describe('does not flush after `beforeEach()`', () => {
        beforeEach(() => __awaiter(void 0, void 0, void 0, function* () {
            yield new Promise((resolve) => {
                setTimeout(() => {
                    resolve();
                });
            });
            // Called because the above Promise is implicitly flushed.
            expect(true).toBe(true);
        }));
        it('empty test so `beforeEach()` will run', () => { });
    });
    describe('does not flush after `afterEach()`', () => {
        afterEach(() => __awaiter(void 0, void 0, void 0, function* () {
            yield new Promise((resolve) => {
                setTimeout(() => {
                    resolve();
                });
            });
            // Called because the above Promise is implicitly flushed.
            expect(true).toBe(true);
        }));
        it('empty test so `afterEach()` will run', () => { });
    });
});
describe('setupModule', () => {
    it('logs unknown element error without CUSTOM_ELEMENTS_SCHEMA', () => {
        const spy = spyOn(console, 'error');
        setupModule({ imports: [TestModule] });
        bootstrapTemplate('<custom-el></custom-el>');
        expect(spy.calls.mostRecent().args[0])
            .toMatch(/'custom-el' is not a known element/);
    });
    it('supports schemas', () => {
        setupModule({ imports: [TestModule], schemas: [CUSTOM_ELEMENTS_SCHEMA] });
        bootstrapTemplate('<custom-el></custom-el>'); // No errors thrown.
    });
});
describe('catalyst with AOT-compiled modules', () => {
    beforeEach(() => {
        setupModule({
            imports: [AotTestModule],
        });
    });
    it('bootstraps', () => {
        bootstrap(AotTestComponent);
        const root = getEl(); // getEl() is the root app element.
        expect(root.textContent).toEqual('Hello World face\n');
    });
    it('bootstraps again', () => {
        bootstrap(AotTestComponent);
        const root = getEl(); // getEl() is the root app element.
        expect(root.textContent).toEqual('Hello World face\n');
    });
    it('can throw before and after bootstrap', () => {
        expect(() => {
            now(() => {
                throw new Error('Error in now before bootstrap');
            });
        }).toThrowError(/Error in now before bootstrap/);
        bootstrap(AotTestComponent);
        const root = getEl(); // getEl() is the root app element.
        expect(root.textContent).toEqual('Hello World face\n');
        expect(() => {
            now(() => {
                throw new Error('Error in now after bootstrap');
            });
        }).toThrowError(/Error in now after bootstrap/);
    });
});
describe('catalyst with AOT-compiled modules', () => {
    beforeEach(() => {
        setupModule({
            imports: [AotTestModule],
            templateOverrides: [{
                    component: AotTestComponent,
                    template: `Template overridden`,
                }]
        });
    });
    it('can override templates', () => {
        bootstrap(AotParentComponent);
        expect(getEl().textContent).toEqual('Template overridden');
    });
    it('can override templates with bootstrapTemplate', () => {
        bootstrapTemplate(`<aot-test-component></aot-test-component>`, {});
        expect(getEl().textContent).toEqual('Template overridden');
    });
});
describe('Automatic event queue processing', () => {
    it('turned off during change detection cycle', () => {
        setupModule({ imports: [TestModule] });
        // Component is able to bootstrap without
        // ExpressionChangedAfterItHasBeenCheckedError.
        bootstrap(TestComponent9);
        expect(getEl().textContent).toBe('Hello World');
    });
    it('turned off in RenderFactory2::end()', () => {
        setupModule({ imports: [TestModule] });
        // Create a RendererFactory2 that triggers events in rf.end(). This can
        // happen, for example, when the animation engine removes an element that
        // has focus in rf.end(), which triggers a focusout event. See b/128703586
        // for details.
        const rf = get(RendererFactory2);
        const origEnd = rf.end;
        rf.end = () => {
            const el = getEl('div');
            if (el) {
                trigger(el, 'click');
            }
            origEnd.apply(rf);
        };
        const comp = bootstrap(TestComponent9);
        // Component is able to update in a microtask without triggering a
        // ExpressionChangedAfterItHasBeenCheckedError.
        comp.updateWithMicrotask();
        flush();
        expect(getEl().textContent).toBe('Hello Universe');
    });
});
describe('afterEach', () => {
    let comp;
    let destroy;
    beforeEach(() => {
        setupModule({ imports: [TestModule] });
    });
    // The following test depends on this one. We could test afterEach
    // behavior with an afterAll, but exceptions in afterAll
    // are not handled in karma jasmine. See
    // https://github.com/karma-runner/karma-jasmine/issues/161
    it('sets up the next test', () => {
        comp = bootstrap(DestroyTestComponentParent);
        destroy = comp.onDestroySpy;
    });
    it('destroys the component after each test', () => {
        expect(destroy).toHaveBeenCalled();
    });
});
describe('beforeEach.async', () => {
    let testVariable;
    beforeEach.async(() => __awaiter(void 0, void 0, void 0, function* () {
        testVariable = 'iAlwaysGetTheShemp';
        const asyncMethod = () => __awaiter(void 0, void 0, void 0, function* () {
            yield null;
            testVariable = 'youGotThatShemp';
        });
        yield asyncMethod();
    }));
    it('runs async tasks before each test', () => {
        expect(testVariable).toEqual('youGotThatShemp');
    });
});
describe('flush', () => {
    beforeEach(() => {
        setupModule({ imports: [TestModule] });
    });
    it('can be called before bootstrap', () => {
        let run = false;
        now(() => {
            setTimeout(() => {
                run = true;
            }, 100);
        });
        expect(run).toBe(false);
        flush();
        expect(run).toBe(true);
    });
    it('handles async tasks scheduled during change detection', () => {
        let resolved = false;
        const instance = {
            name: 'World',
            onResolved: () => {
                resolved = true;
            }
        };
        bootstrapTemplate(`<test-component-4 [name]="name" (resolved)="onResolved()">
       </test-component-4>`, instance);
        instance.name = 'Changed';
        flush();
        expect(getEl().textContent).toBe('Hello Changed');
        expect(resolved).toBe(true);
    });
    it('flushes periodic timers at least once', () => {
        let x = 0, y = 0;
        const start = Date.now();
        now(() => {
            setInterval(() => {
                x++;
            }, 100);
            setInterval(() => {
                y++;
            }, 400);
        });
        flush();
        expect(x).toBe(4);
        expect(y).toBe(1);
        expect(Date.now() - start).toBe(400);
    });
    it('flushes requestAnimationFrame', () => {
        let run = false;
        const start = Date.now();
        let end = 0;
        now(() => {
            requestAnimationFrame(() => {
                run = true;
                end = Date.now();
            });
        });
        flush();
        expect(run).toBe(true);
        expect(end - start).toBe(16);
    });
    it('flushes RxJS scheduler', () => {
        const seen = [];
        const observable = from([1, 2, 3]).pipe(delay(500));
        now(() => {
            observable.subscribe(data => {
                seen.push(data);
            });
        });
        expect(seen).toEqual([]);
        flush();
        expect(seen).toEqual([1, 2, 3]);
    });
    it('flushes RxJS debounceTime', () => {
        let seen = 0;
        now(() => {
            const observable = new Observable((obs) => {
                obs.next(42);
            });
            observable.pipe(debounceTime(100)).subscribe((data) => {
                seen = data;
            });
        });
        expect(seen).toBe(0);
        tick(101);
        expect(seen).toBe(42);
    });
});
describe('markForCheckAndFlush', () => {
    beforeEach(() => {
        setupModule({ imports: [TestModule] });
    });
    it('updates template of component with on push strategy after input changing', () => {
        const comp = bootstrap(TestComponent11);
        // reference type input
        comp.obj = { name: 'Banana' };
        markForCheckAndFlush();
        expect(getEl().textContent).toBe('Name: Banana. Age: 32.');
        // pritimive type input
        comp.primitive = 42;
        markForCheckAndFlush();
        expect(getEl().textContent).toBe('Name: Banana. Age: 42.');
    });
});
describe('bootstrap', () => {
    beforeEach(() => {
        setupModule({ imports: [TestModule] });
    });
    it('can be called after get()', () => {
        get(NameService);
        bootstrap(TestComponent);
        expect(getEl().textContent).toEqual('Hello World face');
    });
    it('binds input values', () => {
        const inputVal = 'banana';
        const comp = bootstrap(TestComponent3, {
            inputVal,
        });
        expect(comp.ngOnInitCalledTimes).toBe(1);
        expect(comp.ngOnInitInputVal).toBe(inputVal);
    });
    it('runs async tasks in onInit', () => {
        const comp = bootstrap(TestComponent5, { name: 'World' });
        expect(comp.resolved).toBe(true);
    });
    it('runs "init" before initial change detection', () => {
        bootstrap(TestComponent6, {}, (component) => {
            spyOn(component, 'getName').and.returnValue('Spy');
        });
        expect(getEl().textContent).toBe('Hello Spy');
    });
    it('does not flush if flush is turned off', () => {
        const comp = bootstrap(TestComponent5, { name: 'World' }, undefined, false /* flush */);
        expect(getEl().textContent).toBe('Hello World');
        expect(comp.resolved).toBe(false);
    });
});
describe('bootstrap from template string', () => {
    beforeEach(() => {
        setupModule({ imports: [TestModule] });
    });
    it('works', () => {
        bootstrapTemplate(`<test-component [name]="name" (click)="name = 'Clicked'">
        </test-component>`, { name: 'Earth' });
        const root = getEl();
        expect(root.textContent).toBe('Hello Earth face');
        getEl('test-component').click();
        expect(root.textContent).toBe('Hello Clicked face');
    });
    it('reflects change in component instance after now()', () => {
        const instance = {
            name: 'Earth',
            click: () => {
                instance.name = 'Clicked';
            }
        };
        bootstrapTemplate('<test-component [name]="name"></test-component>', instance);
        const root = getEl();
        expect(root.textContent).toBe('Hello Earth face');
        now(() => instance.name = 'Venus');
        expect(root.textContent).toBe('Hello Venus face');
        now(() => {
            instance.click();
        });
        expect(root.textContent).toBe('Hello Clicked face');
    });
    it('cannot be called after get()', () => {
        get(NameService);
        expect(() => {
            bootstrapTemplate('<test-component></test-component>', {});
        }).toThrowError('Cannot call bootstrapTemplate() after get()');
    });
    it('runs "init" before initial change detection', () => {
        bootstrapTemplate('<test-component-6></test-component-6>', {}, () => {
            const testComponent = getDebugEl(TestComponent6).componentInstance;
            spyOn(testComponent, 'getName').and.returnValue('Spy');
        });
        expect(getEl().textContent).toBe('Hello Spy');
    });
    it('does not flush if flush is turned off', () => {
        bootstrapTemplate('<test-component-5 name="World"></test-component-5>', {}, undefined, false /* flush */);
        const comp = getDebugEl(TestComponent5).componentInstance;
        expect(getEl().textContent).toBe('Hello World');
        expect(comp.resolved).toBe(false);
    });
    it('allows controller fields to be enumerated', () => {
        bootstrapTemplate('<test-component></test-component>', { foo: 1, bar: 2 });
        const resolved = getDebugEl().componentInstance;
        // Ivy adds an extra '__ngContext__' property so directy check just index 1
        // and 2 so this passes with VE as well.
        expect(Object.keys(resolved)[0]).toEqual('foo');
        expect(Object.keys(resolved)[1]).toEqual('bar');
    });
});
describe('bootstrap with onpush', () => {
    beforeEach(() => {
        setupModule({ imports: [TestModule] });
    });
    it('supports components with onpush change detection', () => {
        const instance = { obj: { name: 'name' } };
        bootstrapTemplate(`<test-component-10 [obj]="obj"></test-component-10>`, instance);
        expect(getEl().textContent).toBe('Name: name');
        now(() => instance.obj.name = 'new name');
        expect(getEl().textContent).not.toBe('Name: new name');
        now(() => instance.obj = { name: 'third name' });
        expect(getEl().textContent).toBe('Name: third name');
    });
});
describe('finding els', () => {
    beforeEach(() => {
        setupModule({ imports: [TestModule] });
        bootstrapTemplate(`<test-component name="foo"></test-component>
       <test-component name="bar"></test-component>
       <a href="http://google.com/">google</a>
       <input type="text" value="val">`, { a: 10 });
    });
    it('returns whether an element is present', () => {
        expect(hasEl('p')).toBe(true);
        expect(hasEl('header')).toBe(false);
    });
    it('finds element by CSS', () => {
        expect(getEl('p')).not.toBe(null);
    });
    it('getEl() throws an exception if element is not found', () => {
        expect(() => getEl('header'))
            .toThrowError('Element not found for "header"');
    });
    it('getEl() returns proper subtypes for simple selectors', () => {
        expect(getEl('a').href).toEqual('http://google.com/');
        expect(getEl('input').value).toEqual('val');
    });
    it('finds all elements queried by CSS', () => {
        const els = getEls('p');
        expect(els.length).toBe(2);
        expect(els[0].textContent).toBe('Hello foo face');
        expect(els[1].textContent).toBe('Hello bar face');
    });
    it('finds a debug element by type', () => {
        const debug = getDebugEl(TestComponent);
        expect(debug.componentInstance.name).toEqual('foo');
    });
    it('finds a debug element by CSS', () => {
        const debug = getDebugEl('[name="bar"]');
        expect(debug.componentInstance.name).toEqual('bar');
    });
    it('find root debug element', () => {
        expect(getDebugEl().componentInstance.a).toEqual(10);
    });
    it('finds debug elements by type', () => {
        const debugs = getDebugEls(TestComponent);
        expect(debugs.length).toEqual(2);
        expect(debugs[0].componentInstance.name).toEqual('foo');
        expect(debugs[1].componentInstance.name).toEqual('bar');
    });
    it('finds debug elements by CSS', () => {
        const debugs = getDebugEls('test-component');
        expect(debugs.length).toEqual(2);
        expect(debugs[0].componentInstance.name).toEqual('foo');
        expect(debugs[1].componentInstance.name).toEqual('bar');
    });
    it('locates debug elements', () => {
        expect(hasDebugEl(TestComponent)).toBe(true);
        expect(hasDebugEl('test-component')).toBe(true);
        expect(hasDebugEl(TestComponent2)).toBe(false);
        expect(hasDebugEl('test-component-2')).toBe(false);
    });
    it('getDebugEl() throws an exception if element is not found', () => {
        expect(() => getDebugEl(TestComponent2))
            .toThrowError(`DebugElement not found for "${TestComponent2}"`);
    });
});
describe('finding els in given root', () => {
    let root;
    let rootDebugElement;
    beforeEach(() => {
        setupModule({ imports: [TestModule] });
        bootstrapTemplate(`
       <span id="first"><div>foo0</div><div>foo1</div></span>
       <span id="second"><p>bar0</p><p>bar1</p></span>`, {});
        root = getEl('span#second');
        rootDebugElement = getDebugEl('span#second');
    });
    it('returns whether an element is present', () => {
        expect(hasEl('p', root)).toBe(true);
        expect(hasEl('div', root)).toBe(false);
    });
    it('finds element by CSS', () => {
        const el = getEl('p', root);
        expect(el).not.toBe(null);
        expect(el.textContent).toBe('bar0');
    });
    it('getEl() throws an exception if element is not found', () => {
        expect(() => getEl('div', root))
            .toThrowError('Element not found for "div"');
    });
    it('finds all elements queried by CSS', () => {
        const els = getEls('p', root);
        expect(els.length).toBe(2);
        expect(els[0].textContent).toBe('bar0');
        expect(els[1].textContent).toBe('bar1');
    });
    it('returns whether a debug element is present', () => {
        expect(hasDebugEl('p', rootDebugElement)).toBe(true);
        expect(hasDebugEl('div', rootDebugElement)).toBe(false);
    });
    it('finds debug element by CSS', () => {
        const el = getDebugEl('p', rootDebugElement);
        expect(el).not.toBe(null);
        expect(el.nativeElement.textContent).toBe('bar0');
    });
    it('getDebugEl() throws an exception if element is not found', () => {
        expect(() => getDebugEl('div', rootDebugElement))
            .toThrowError('DebugElement not found for "div"');
    });
    it('finds all debug elements queried by CSS', () => {
        const els = getDebugEls('p', rootDebugElement);
        expect(els.length).toBe(2);
        expect(els[0].nativeElement.textContent).toBe('bar0');
        expect(els[1].nativeElement.textContent).toBe('bar1');
    });
});
const mockedTestDate = new Date('22 Jan 1983 12:00:00 GMT');
describeWithDate('task control', mockedTestDate, () => {
    let root;
    beforeEach(() => {
        expect(new Date()).toEqual(mockedTestDate);
        setupModule({ imports: [TestModule] });
        bootstrap(TestComponent2);
        root = getEl();
    });
    it('tick advances mock Date', () => {
        tick(1000);
        expect(Date.now()).toBe(new Date('22 Jan 1983 12:00:01 GMT').getTime());
    });
    it('tick works with RxJS scheduler', () => {
        const seen = [];
        const observable = from([1, 2, 3]).pipe(delay(500));
        now(() => {
            observable.subscribe(data => {
                seen.push(data);
            });
        });
        expect(seen).toEqual([]);
        tick(500);
        expect(seen).toEqual([1, 2, 3]);
    });
    it('tick can advance clock to trigger timers', () => {
        expect(root.textContent).toEqual('Hello World');
        getEl('p').click();
        expect(root.textContent).toEqual('Hello Clicked');
        tick(1000);
        expect(root.textContent).toEqual('Hello Timeout');
    });
    it('flush can trigger all queued timers', () => {
        expect(root.textContent).toEqual('Hello World');
        getEl('p').click();
        expect(root.textContent).toEqual('Hello Clicked');
        flush();
        expect(root.textContent).toEqual('Hello Timeout');
    });
});
describe('changing the test module', () => {
    it('is able to declare components via setupModule', () => {
        setupModule({ imports: [MatIconModule], declarations: [TestComponent] });
        bootstrap(TestComponent);
        expect(getEl().textContent).toEqual('Hello World face');
    });
    it('is able to declare regular providers via setupModule', () => {
        const mockName = { name: 'mock name' };
        setupModule({
            imports: [TestModule],
            providers: [{ provide: NameService, useValue: mockName }],
        });
        bootstrap(TestComponent7);
        // Getting through the top-level injector should return the mocked one.
        expect(get(NameService).name).toBe('mock name');
        // Regular providers cannot override Component level providers
        expect(getEl().textContent).toBe('Hello original');
    });
    it('is able to add multiproviders via setupModule', () => {
        setupModule({
            imports: [TestModule],
            providers: [
                { provide: MULTI_NAME_SERVICE, useValue: { name: 'Name2' }, multi: true }
            ],
        });
        bootstrap(TestComponent8);
        expect(getEl().textContent).toBe('Hello Name1 Name2');
    });
    it('overrides providers from anywhere', () => {
        const mockName = { name: 'mock name' };
        setupModule({
            imports: [TestModule],
            superProviders: [{ provide: NameService, useValue: mockName }]
        });
        const component = bootstrap(TestComponent7);
        expect(component.getName()).toEqual('mock name');
        // Make sure that useValue uses the exact object.
        expect(component.nameService).toBe(mockName);
    });
    it('overrides providers from anywhere when providers is an array', () => {
        const mockName = { name: 'mock name' };
        setupModule({
            imports: [TestModule],
            superProviders: [[{ provide: NameService, useValue: mockName }]]
        });
        const component = bootstrap(TestComponent7);
        expect(component.getName()).toEqual('mock name');
        // Make sure that useValue uses the exact object.
        expect(component.nameService).toBe(mockName);
    });
    it('overrides providers not provided by any NgModule', () => {
        const mockName = { name: 'mock name' };
        setupModule({
            imports: [TestModule],
            superProviders: [{ provide: UnProvidedService, useValue: mockName }]
        });
        const service = get(UnProvidedService);
        expect(service.name).toEqual('mock name');
        // Make sure that useValue uses the exact object.
        expect(service).toBe(mockName);
    });
});
describe('helper functions', () => {
    it('triggers event', () => {
        setupModule({ imports: [TestModule] });
        const spyHandler = jasmine.createSpy('handler');
        bootstrapTemplate('<input>', {});
        const input = getEl('input');
        input.addEventListener('focus', spyHandler);
        trigger(input, 'focus');
        expect(spyHandler).toHaveBeenCalled();
        // Remove event listener so the node could be fully removed from memory
        input.removeEventListener('focus', spyHandler);
    });
    it('triggers a keyboard event without parameter', () => {
        setupModule({ imports: [TestModule] });
        const spyHandler = jasmine.createSpy('handler');
        bootstrapTemplate('<div>', {});
        const div = getEl('div');
        div.addEventListener('keypress', spyHandler);
        trigger(div, 'keypress');
        expect(spyHandler).toHaveBeenCalled();
        // Remove event listener so the node could be fully removed from memory
        div.removeEventListener('keypress', spyHandler);
    });
    it('triggers a keyboard event with a keyCode', () => {
        setupModule({ imports: [TestModule] });
        const spyHandler = jasmine.createSpy('handler');
        bootstrapTemplate('<div>', {});
        const div = getEl('div');
        div.addEventListener('keypress', spyHandler);
        trigger(div, 'keypress', { code: 'ENTER' }, { 'keyCode': 13 });
        expect(spyHandler)
            .toHaveBeenCalledWith(jasmine.objectContaining({ code: 'ENTER', keyCode: 13 }));
        // Remove event listener so the node could be fully removed from memory
        div.removeEventListener('keypress', spyHandler);
    });
});
describe('service testing', () => {
    it('can inject a service without bootstrapping a component', () => {
        const mockName = { name: 'mock name' };
        setupModule({
            imports: [TestModule],
            superProviders: [{ provide: NameService, useValue: mockName }]
        });
        const service = get(NameService);
        expect(service.name).toEqual('mock name');
        // Make sure that useValue uses the exact object.
        expect(service).toBe(mockName);
    });
});
describe('typed blocks', () => {
    beforeEach(function () {
        expect(this.a).toBeUndefined();
        this.a = 10;
    });
    it('work', function () {
        expect(this.a).toBe(10);
    });
    it('work again with context cleared', function () {
        expect(this.a).toBeGreaterThan(9);
    });
});
describe('SVG elements', () => {
    beforeEach(() => {
        setupModule({
            imports: [TestModule],
        });
        bootstrap(SvgTestComponent);
    });
    it('getEl can get an SVG element', () => {
        const svg = getEl('svg');
        expect(svg.id).toEqual('1');
        const circle = getEl('circle', svg);
        expect(circle.id).toEqual('circle');
    });
    it('getEls can get SVG elements', () => {
        const svgs = getEls('svg');
        expect(svgs.length).toEqual(2);
        expect(svgs[0].id).toEqual('1');
        expect(svgs[1].id).toEqual('2');
    });
    it('getDebugEl can get an SVG element', () => {
        const svg = getDebugEl('svg').nativeElement;
        expect(svg.id).toEqual('1');
        const circle = getDebugEl('circle', getDebugEl('svg')).nativeElement;
        expect(circle.id).toEqual('circle');
    });
    it('getDebugEls can get SVG elements', () => {
        const svgs = getDebugEls('svg');
        expect(svgs.length).toEqual(2);
        expect(svgs[0].nativeElement.id).toEqual('1');
        expect(svgs[1].nativeElement.id).toEqual('2');
    });
});
describe('destroyTestComponent', () => {
    beforeEach(() => {
        setupModule({
            imports: [TestModule],
        });
        bootstrapTemplate(`<destroy-test-component-parent>
                      </destroy-test-component-parent>
                      <destroy-test-component-child>
                      </destroy-test-component-child>`);
    });
    it('causes destroys all children components', () => {
        const parentOnDestroy = getDebugEl(DestroyTestComponentParent).componentInstance.onDestroySpy;
        expect(parentOnDestroy).toHaveBeenCalledTimes(0);
        const childOnDestroys = getDebugEls(DestroyTestComponentChild)
            .map(child => child.componentInstance.onDestroySpy);
        expect(childOnDestroys.length).toBe(2);
        expect(childOnDestroys[0]).toHaveBeenCalledTimes(0);
        expect(childOnDestroys[1]).toHaveBeenCalledTimes(0);
        destroyTestComponent();
        expect(parentOnDestroy).toHaveBeenCalledTimes(1);
        expect(childOnDestroys[0]).toHaveBeenCalledTimes(1);
        expect(childOnDestroys[1]).toHaveBeenCalledTimes(1);
        // Second destroy does fails.
        expect(() => {
            destroyTestComponent();
        }).toThrowError('Please call bootstrap() first');
    });
});
//----------- Components under test. Usually in a different file.--------------
let TestComponent = class TestComponent {
    constructor() {
        this.name = 'World';
    }
    click() {
        Promise.resolve(null).then(() => null).then(() => this.name = 'Clicked');
    }
    ngOnDestroy() { }
};
__decorate([
    Input()
], TestComponent.prototype, "name", void 0);
TestComponent = __decorate([
    Component({
        preserveWhitespaces: true,
        selector: 'test-component',
        template: `<p (click)="click()">Hello {{name}} <mat-icon>face</mat-icon></p>`,
    })
], TestComponent);
let TestComponent2 = class TestComponent2 {
    constructor() {
        this.name = 'World';
    }
    click() {
        this.name = 'Clicked';
        setTimeout(() => Promise.resolve(null).then(() => this.name = 'Timeout'), 1000);
    }
};
TestComponent2 = __decorate([
    Component({
        preserveWhitespaces: true,
        selector: 'test-component-2',
        template: '<p (click)="click()">Hello {{name}}</p>',
    })
], TestComponent2);
let TestComponent3 = class TestComponent3 {
    constructor() {
        this.ngOnInitCalledTimes = 0;
    }
    ngOnInit() {
        this.ngOnInitCalledTimes++;
        this.ngOnInitInputVal = this.inputVal;
    }
};
__decorate([
    Input()
], TestComponent3.prototype, "inputVal", void 0);
TestComponent3 = __decorate([
    Component({
        preserveWhitespaces: true,
        selector: 'test-component-3',
        template: '<div></div>',
    })
], TestComponent3);
let TestComponent4 = class TestComponent4 {
    constructor() {
        this.resolved = new EventEmitter();
    }
    ngOnChanges() {
        Promise.resolve().then(() => {
            this.resolved.emit(true);
        });
    }
};
__decorate([
    Input()
], TestComponent4.prototype, "name", void 0);
__decorate([
    Output()
], TestComponent4.prototype, "resolved", void 0);
TestComponent4 = __decorate([
    Component({
        preserveWhitespaces: true,
        selector: 'test-component-4',
        template: '<div>Hello {{name}}</div>',
    })
], TestComponent4);
let TestComponent5 = class TestComponent5 {
    constructor() {
        this.resolved = false;
    }
    ngOnInit() {
        Promise.resolve().then(() => {
            this.resolved = true;
        });
    }
};
__decorate([
    Input()
], TestComponent5.prototype, "name", void 0);
TestComponent5 = __decorate([
    Component({
        preserveWhitespaces: true,
        selector: 'test-component-5',
        template: '<div>Hello {{name}}</div>',
    })
], TestComponent5);
let TestComponent6 = class TestComponent6 {
    getName() {
        return 'World';
    }
};
TestComponent6 = __decorate([
    Component({
        preserveWhitespaces: true,
        selector: 'test-component-6',
        template: '<div>Hello {{getName()}}</div>',
    })
], TestComponent6);
let UnProvidedService = class UnProvidedService {
    constructor() {
        this.name = 'original';
    }
};
UnProvidedService = __decorate([
    Injectable()
], UnProvidedService);
let NameService = class NameService {
    constructor() {
        this.name = 'original';
    }
};
NameService = __decorate([
    Injectable()
], NameService);
let MultiNameService = class MultiNameService {
};
MultiNameService = __decorate([
    Injectable()
], MultiNameService);
const MULTI_NAME_SERVICE = new InjectionToken('MultiNameService');
let TestComponent7 = class TestComponent7 {
    constructor(nameService) {
        this.nameService = nameService;
    }
    getName() {
        return this.nameService.name;
    }
};
TestComponent7 = __decorate([
    Component({
        preserveWhitespaces: true,
        selector: 'test-component-7',
        template: '<div>Hello {{getName()}}</div>',
        providers: [NameService],
    })
], TestComponent7);
let TestComponent8 = class TestComponent8 {
    constructor(nameServices) {
        this.nameServices = nameServices;
    }
    getName() {
        let name = '';
        this.nameServices.forEach(s => {
            name += ' ' + s.name;
        });
        return name;
    }
};
TestComponent8 = __decorate([
    Component({
        preserveWhitespaces: true,
        selector: 'test-component-8',
        template: '<div>Hello{{getName()}}</div>',
    }),
    __param(0, Inject(MULTI_NAME_SERVICE))
], TestComponent8);
let TestComponent9 = class TestComponent9 {
    constructor() {
        this.name = '';
    }
    ngAfterViewInit() {
        // Changing name as part of change detection, but in the next tick should
        // be ok.
        Promise.resolve(null).then(() => {
            this.name = 'World';
        });
        // Trigger an event handler during the change detection. This should
        // not cause the above Promise to be processed in the same tick.
        this.root.nativeElement.click();
    }
    clicked() {
        // empty.
    }
    // Queue up a microtask that changes a binding. This should never cause an
    // ExpressionChangedAfterItHasBeenCheckedError
    updateWithMicrotask() {
        Promise.resolve(null).then(() => {
            this.name = 'Universe';
        });
    }
};
__decorate([
    ViewChild('root', { static: false })
], TestComponent9.prototype, "root", void 0);
TestComponent9 = __decorate([
    Component({
        preserveWhitespaces: true,
        selector: 'test-component-9',
        template: '<div #root (click)="clicked()">Hello {{name}}</div>',
    })
], TestComponent9);
let TestComponent10 = class TestComponent10 {
    constructor() {
        this.obj = {};
    }
};
__decorate([
    Input()
], TestComponent10.prototype, "obj", void 0);
TestComponent10 = __decorate([
    Component({
        changeDetection: ChangeDetectionStrategy.OnPush,
        preserveWhitespaces: true,
        selector: 'test-component-10',
        template: `Name: {{obj.name}}`,
    })
], TestComponent10);
let TestComponent11 = class TestComponent11 {
    constructor() {
        this.obj = {};
        this.primitive = 32;
    }
};
__decorate([
    Input()
], TestComponent11.prototype, "obj", void 0);
__decorate([
    Input()
], TestComponent11.prototype, "primitive", void 0);
TestComponent11 = __decorate([
    Component({
        changeDetection: ChangeDetectionStrategy.OnPush,
        selector: 'test-component-11',
        template: `Name: {{obj.name}}. Age: {{primitive}}.`,
    })
], TestComponent11);
let SvgTestComponent = class SvgTestComponent {
};
SvgTestComponent = __decorate([
    Component({
        selector: 'svg-test-component',
        template: `<svg id="1"><circle id="circle"></circle></svg><svg id="2"></svg>`,
    })
], SvgTestComponent);
let DestroyTestComponentParent = class DestroyTestComponentParent {
    constructor() {
        this.onDestroySpy = jasmine.createSpy('DestroyTestComponentParent#onDestroy');
    }
    ngOnDestroy() {
        this.onDestroySpy();
    }
};
DestroyTestComponentParent = __decorate([
    Component({
        selector: 'destroy-test-component-parent',
        template: '<destroy-test-component-child></destroy-test-component-child>'
    })
], DestroyTestComponentParent);
let DestroyTestComponentChild = class DestroyTestComponentChild {
    constructor() {
        this.onDestroySpy = jasmine.createSpy('DestroyTestComponentChild#onDestroy');
    }
    ngOnDestroy() {
        this.onDestroySpy();
    }
};
DestroyTestComponentChild = __decorate([
    Component({ selector: 'destroy-test-component-child', template: '' })
], DestroyTestComponentChild);
let TestModule = class TestModule {
};
TestModule = __decorate([
    NgModule({
        declarations: [
            TestComponent,
            TestComponent2,
            TestComponent3,
            TestComponent4,
            TestComponent5,
            TestComponent6,
            TestComponent7,
            TestComponent8,
            TestComponent9,
            TestComponent10,
            TestComponent11,
            SvgTestComponent,
            DestroyTestComponentParent,
            DestroyTestComponentChild,
        ],
        imports: [MatIconModule],
        providers: [
            NameService,
            { provide: MULTI_NAME_SERVICE, useValue: { name: 'Name1' }, multi: true },
        ],
        exports: [
            TestComponent,
            TestComponent4,
            TestComponent5,
            TestComponent6,
            TestComponent7,
            TestComponent8,
            TestComponent9,
            TestComponent10,
            TestComponent11,
            SvgTestComponent,
            DestroyTestComponentParent,
            DestroyTestComponentChild,
        ],
    })
], TestModule);
