// @angular/catalyst/dist/dist/shared-angular-catalyst/test_component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Component, Input, NgModule } from '@angular/core';
import { MatIconModule } from '@angular/material/icon';
let AotTestComponent = class AotTestComponent {
    constructor() {
        this.name = 'World';
    }
    click() { }
};
__decorate([
    Input()
], AotTestComponent.prototype, "name", void 0);
AotTestComponent = __decorate([
    Component({
        preserveWhitespaces: true,
        selector: 'aot-test-component',
        templateUrl: './test_component.ng.html',
    })
], AotTestComponent);
export { AotTestComponent };
let AotParentComponent = class AotParentComponent {
};
AotParentComponent = __decorate([
    Component({
        preserveWhitespaces: true,
        selector: 'aot-parent-test-component',
        template: `<aot-test-component name="World"></aot-test-component>`,
    })
], AotParentComponent);
export { AotParentComponent };
let AotTestModule = class AotTestModule {
};
AotTestModule = __decorate([
    NgModule({
        imports: [MatIconModule],
        declarations: [AotParentComponent, AotTestComponent],
        exports: [AotTestComponent],
    })
], AotTestModule);
export { AotTestModule };
